### PR TITLE
DevTools: Improve browser extension iframe support

### LINKED ISSUE #18945: DevTools: Improve browser extension iframe support
When react is inside an iframe, chrome extension for react devtools fails to detect react.

This is because the extension sets `__REACT_DEVTOOLS_GLOBAL_HOOK__` only on the top level window. Apparently it's possible to have `__REACT_DEVTOOLS_GLOBAL_HOOK__` on iframes too by adding `all_frames: true` in extension [manifest.json](https://github.com/facebook/react/blob/c390ab3643612dc08ca4bebadc5b0377e9e7eb79/packages/react-devtools-extensions/chrome/manifest.json#L49). It was done by redux devtools extension in https://github.com/zalmoxisus/redux-devtools-extension/pull/56.

Have you considered adding `all_frames: true` to chrome extension ?

## Steps To Reproduce

1. Create a file `react-main.html`
```html
<!DOCTYPE html>
<html>
  <head>
    <script crossorigin src="https://unpkg.com/react@16/umd/react.development.js"></script>
    <script crossorigin src="https://unpkg.com/react-dom@16/umd/react-dom.development.js"></script>
  </head>
  <body></body>
</html>

```
2. Create a file `react-iframe.html`
```html
<!DOCTYPE html>
<html>
  <head></head>
  <body>
    <iframe src="./react-main.html" />
  </body>
</html>
```
3. Open `react-iframe.html` in chrome

## The current behavior

react-devtools-not-detected

## The expected behavior

react-devtools-detected

Pull request: https://github.com/facebook/react/pull/18952

### PR DESCRIPTION
**Pull request by omarsy**
Fixes #18945
## Summary
Correct the issue caused by this PR https://github.com/facebook/react/pull/19827

### CODE REVIEW

File: packages/react-devtools-shared/src/hook.js [hunk 1]
```diff
@@ -302,7 +302,10 @@ export function installHook(target: any): DevToolsHook | null {
   const renderers = new Map();
 
   let hook: DevToolsHook;
-  if (!canAccessParentWindow(target) || isMainWindow(target)) {
+  if (canAccessParentWindow(target) && !isMainWindow(target)) {
+    hook = getMainWindow(target).__REACT_DEVTOOLS_GLOBAL_HOOK__;
+  }
+  if (!hook) {
```
Comment (omarsy) [inline packages/react-devtools-shared/src/hook.js:308]: I handle know if we don't have `__REACT_DEVTOOLS_GLOBAL_HOOK__` in the parent window

**Review by bvaughn — APPROVED** [review]
Looks like these changes fix the dev shell regression 👍 

So I guess they're good to move forward with.

**Comment by omarsy** [comment]
Thank you @bvaughn !

**Comment by StanislavMayorov** [comment]
@omarsy I've built the extension from master `97625272a`, but react devtools doesn't work with iframe. Is it required any extra configuration in my project? 
I use different origins for iframe and main app. Web app is started from `localhost:3333` and iframe app from `localhost:3001`.

**Comment by omarsy** [comment]
@StanislavMayorov No you need any extra configuration. Can I see your code  ? What kind of iframe do you use (sandbox or not) ?

**Comment by omarsy** [comment]
Your previous devtool work  with iframe if you add  if you add this code in your iframe
```
<script type="text/javascript">
      // Enable DevTools to inspect React inside of an <iframe>	
      // This must run before React is loaded	
      __REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;	
    </script>
``` ?

**Comment by StanislavMayorov** [comment]
I don't have a demo that I can share right now. That's sandbox attr.
```
<iframe
   sandbox="allow-forms allow-same-origin allow-scripts allow-popups"
 />
```

**Comment by StanislavMayorov** [comment]
> Your previous devtool work with iframe if you add if you add this code in your iframe
> 
> ```
> <script type="text/javascript">
>       // Enable DevTools to inspect React inside of an <iframe>	
>       // This must run before React is loaded	
>       __REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;	
>     </script>
> ``` ?
> ```

No, It didn't work for my app.

**Comment by omarsy** [comment]
In the case of the sandbox we cannot always access the parent's window (blocked by the browser for security reasons), that's why it doesn't work. My fix allows this script 
```
<script type="text/javascript">
      // Enable DevTools to inspect React inside of an <iframe>	
      // This must run before React is loaded	
      __REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;	
    </script>
```
to be done automatically. But this does not solve the problem of the sandbox. Maybe we can solve the case of sandboxed iframes but I think it is necessary to make a big refactoring and at the security level it remains a big question mark.

**Comment by StanislavMayorov** [comment]
I think it doesn't have access to `window.parent.__REACT_DEVTOOLS_GLOBAL_HOOK__`. But redux dev works fine.

The extension (version `97625272a`) crashes If I remove sandbox attribute.
Screenshot 2020-10-05 at 13 51 14

**Comment by StanislavMayorov** [comment]
It doesn't crash if I run chrome without CORS using `--args --user-data-dir="/tmp/someFolderName" --disable-web-security`

**Comment by bvaughn** [comment]
Sounds like maybe this commit needs to be reverted again?

**Comment by omarsy** [comment]
I don't know how to reproduce this issue. Maybe if I have the full error it can help me. Normally I added this function 
```
function canAccessParentWindow(targetWindow: any): boolean {
    try {
      return !!targetWindow.parent.origin;
    } catch (error) {
      return false;
    }
  }
``` to protect us for this kind issue

**Comment by StanislavMayorov** [comment]
Did you try to use iframe and app with different origins to reproduce?
I will try to create a demo soon.

**Comment by omarsy** [comment]
Yes, I try it. Maybe I have wrong http headers ? Ok thank you @StanislavMayorov

File: fixtures/devtools/iframe/iframe-same-origin-sandbox.html [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<!DOCTYPE html>
```
Comment (omarsy) [inline fixtures/devtools/iframe/iframe-same-origin-sandbox.html:1]: Maybe you can add your case like this and push a PR ?

Comment (bvaughn) [reply fixtures/devtools/iframe/iframe-same-origin-sandbox.html:1]: This is a good suggestion!

**Comment by StanislavMayorov** [comment]
I tried to reproduce the crash using a demo app, but it works okay with different origins. I couldn't find out what is the reason of crash in my app. I tested that `sandbox` attr doesn't affect the crash in my app.

Screenshot 2020-10-07 at 23 54 42
Screenshot 2020-10-07 at 23 55 00
Screenshot 2020-10-07 at 23 55 12

**Comment by omarsy** [comment]
Perhaps you have an iframe element in your props ? In this case, when we try to dehydrate the props, it causes this issue. 
If you add a debbuger in this line you can see which props causes the issue
https://github.com/facebook/react/blob/4ead6b53057ee6c6129a6d2f6e264232130b1fce/packages/react-devtools-shared/src/utils.js#L614

**Comment by omarsy** [comment]
@bvaughn I think it’s not related to my merge. I can reproduce it . 
https://codesandbox.io/s/modest-moser-jg9x9?file=/src/index.js

**Comment by StanislavMayorov** [comment]
I don't have crash in this demo.
> @bvaughn I think it’s not related to my merge. I can reproduce it.
> https://codesandbox.io/s/modest-moser-jg9x9?file=/src/index.js

I'll check props as I know there is no element.

**Comment by omarsy** [comment]
> I don't have crash in this demo.
Try to inspect the App

**Comment by StanislavMayorov** [comment]
@omarsy I just open your app and see error, even without react devtools.

But I have following steps in my app:
- open app
- inspect react component

result: it fails with error.

**Comment by omarsy** [comment]
Yes, normally if it was due to my change it would have crashed before the inspect (because we do several treatments with the window).  I think you have a window in one of your props. I think i can have a PR for this issue

**Comment by bvaughn** [comment]
If you are confident that the crash is not related to the previous iframe change, then you can submit a PR that reverts my revert.

**Comment by StanislavMayorov** [comment]
I can check current version with reverted PR.

**Comment by bvaughn** [comment]
@StanislavMayorov You good to build and test this locally? (Let me know if you need me to build and attach a version here.)

**Comment by omarsy** [comment]
With the reverted PR you cannot inspect 'iframe react component' I think.  We need to give you a quick fix for the window security crash + iframe support

**Comment by bvaughn** [comment]
Couldn't you add the pass through (like in our fixtures) to test with the reverted DT?

**Comment by omarsy** [comment]
@bvaughn  for this issue https://codesandbox.io/s/modest-moser-jg9x9?file=/src/index.js. Can I push a PR ?

**Comment by bvaughn** [comment]
Sorry @omarsy but I don't have time to dig into this at the moment. Please feel free to submit a PR if you think it's appropriate :)

**Comment by StanislavMayorov** [comment]
> Couldn't you add the pass through (like in our fixtures) to test with the reverted DT?
> 
>https://github.com/facebook/react/blob/6eca8eff08e293229c2ff957f175573faf994fc0/fixtures/devtools/regression/16.7.html#L8-L12

Okay, I have this changes in my app. I will test it. Should I use this [version](https://github.com/facebook/react/commit/461cd84944c28a5715a16a19f2503c87068832d6)? 



> @StanislavMayorov You good to build and test this locally? (Let me know if you need me to build and attach a version here.)

I have already built it using these steps https://github.com/facebook/react/tree/master/packages/react-devtools-extensions#build-from-source

**Comment by omarsy** [comment]
> Okay, I have this changes in my app. I will test it. Should I use this version?

Yes I think you can use this version

**Comment by StanislavMayorov** [comment]
I tested this [version](https://github.com/facebook/react/commit/461cd84944c28a5715a16a19f2503c87068832d6) and added `__REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;` in html. 
There's no error and react inspector doesn't work. Inspector works fine if I test with disabled CORS in chrome. 
So I think this PR is the reason of crash.

**Comment by omarsy** [comment]
`__REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;` react inspector should work if you add this.

**Comment by StanislavMayorov** [comment]
> `__REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;` react inspector should work if you add this.

Iframe app don't have access to `parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;`
> I use different origins for iframe and main app. Web app is started from `localhost:3333` and iframe app from `localhost:3001`.

**Comment by omarsy** [comment]
What you say :
this `__REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;`  causes crash in start and with my version it causes a crash when yoy try to inspect the iframe ?

**Comment by omarsy** [comment]
> So I think this PR is the reason of crash.

What I'm saying is in the stacktrace, you can see that we have the issue when trying to desyhdrate props. So my understanding is that you should have a window element in your props because of this but maybe I'm wrong that's why I asked you to put in a debugger to see which props were causing the issue `If you add a debbuger in this line you can see which props causes the issue. `. 

Well I'll stop investigating as I can see I'm not getting anywhere. 

image

**Comment by StanislavMayorov** [comment]
> What you say :
> this `__REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;` causes crash in start and with my version it causes a crash when you try to inspect the iframe ?

You are right. There are two errors. I missed the error on start in console because react-error-overlay doesn't work at this stage. 

The fix adds script `__REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;`, but it causes error in case of different origins, doesn't it? That's not good for default behavior.

**Comment by StanislavMayorov** [comment]
I will try to reproduce and debug props soon. Is it enough just to use `build:dev`?

**Comment by bvaughn** [comment]
> Is it enough just to use `build:dev`?

That's fine. If you're testing in Chrome, the fastest way would be to run either `"build:chrome"` or `"build:chrome:dev"`, e.g.

```sh
yarn build:chrome && yarn test:chrome --url=<url-of-your-sandbox>
```

**Pull request merged**